File: app/javascript/modules/transaction_config.js
/**
 * Transaction Configuration Module
 * 
 * This module provides JavaScript configuration objects for each transaction type,
 * including validation rules, UI behavior, and API endpoints.
 * 
 * These configurations are designed to be passed from Rails views to JavaScript
 * controllers to enable configuration-driven behavior.
 */

export const TRANSACTION_CONFIGS = {
  stock_in: {
    type: 'stock_in',
    title: 'Stock In',
    color: 'green',
    locations: ['destination'],
    validation_rules: ['positive_quantity'],
    quantity_behavior: 'positive',
    description: 'Add items to inventory at a specific location',
    ui_behavior: {
      show_current_stock: true,
      allow_negative_quantity: false,
      require_location_selection: true,
      default_quantity: 1,
      quantity_step: 1,
      quantity_min: 1
    },
    api_endpoints: {
      search: '/teams/{teamId}/items/search',
      transaction: '/teams/{teamId}/transactions'
    },
    validation_messages: {
      no_location: 'Please select a location',
      no_items: 'Please add items and quantities',
      invalid_quantity: 'Quantity must be positive',
      insufficient_stock: 'Not enough stock available'
    },
    css_classes: {
      primary_color: 'green-600',
      hover_color: 'green-700',
      border_color: 'green-200',
      bg_color: 'green-50',
      text_color: 'green-600'
    }
  },

  stock_out: {
    type: 'stock_out',
    title: 'Stock Out',
    color: 'red',
    locations: ['source'],
    validation_rules: ['positive_quantity', 'stock_availability'],
    quantity_behavior: 'positive',
    description: 'Remove items from inventory at a specific location',
    ui_behavior: {
      show_current_stock: true,
      allow_negative_quantity: false,
      require_location_selection: true,
      default_quantity: 1,
      quantity_step: 1,
      quantity_min: 1,
      validate_stock_availability: true
    },
    api_endpoints: {
      search: '/teams/{teamId}/items/search',
      transaction: '/teams/{teamId}/transactions'
    },
    validation_messages: {
      no_location: 'Please select a location',
      no_items: 'Please add items and quantities',
      invalid_quantity: 'Quantity must be positive',
      insufficient_stock: 'Not enough stock available for this item'
    },
    css_classes: {
      primary_color: 'red-600',
      hover_color: 'red-700',
      border_color: 'red-200',
      bg_color: 'red-50',
      text_color: 'red-600'
    }
  },

  adjust: {
    type: 'adjust',
    title: 'Adjust Stock',
    color: 'blue',
    locations: ['destination'],
    validation_rules: ['adjustment_calculation'],
    quantity_behavior: 'adjustment',
    description: 'Adjust item quantities to match actual counts',
    ui_behavior: {
      show_current_stock: true,
      allow_negative_quantity: true,
      require_location_selection: true,
      default_quantity: 0, // Will be set to current stock
      quantity_step: 1,
      show_adjustment_difference: true
    },
    api_endpoints: {
      search: '/teams/{teamId}/items/search',
      transaction: '/teams/{teamId}/transactions'
    },
    validation_messages: {
      no_location: 'Please select a location',
      no_items: 'Please add items and quantities',
      invalid_quantity: 'Please enter a valid quantity',
      no_adjustment: 'No adjustment needed - quantity matches current stock'
    },
    css_classes: {
      primary_color: 'blue-600',
      hover_color: 'blue-700',
      border_color: 'blue-200',
      bg_color: 'blue-50',
      text_color: 'blue-600'
    }
  },

  move: {
    type: 'move',
    title: 'Move Stock',
    color: 'purple',
    locations: ['source', 'destination'],
    validation_rules: ['positive_quantity', 'stock_availability'],
    quantity_behavior: 'positive',
    description: 'Transfer items between locations',
    ui_behavior: {
      show_current_stock: true,
      allow_negative_quantity: false,
      require_location_selection: true,
      require_different_locations: true,
      default_quantity: 1,
      quantity_step: 1,
      quantity_min: 1,
      validate_stock_availability: true
    },
    api_endpoints: {
      search: '/teams/{teamId}/items/search',
      transaction: '/teams/{teamId}/transactions/move'
    },
    validation_messages: {
      no_source_location: 'Please select a source location',
      no_destination_location: 'Please select a destination location',
      same_locations: 'Source and destination locations must be different',
      no_items: 'Please add items and quantities',
      invalid_quantity: 'Quantity must be positive',
      insufficient_stock: 'Not enough stock available at source location'
    },
    css_classes: {
      primary_color: 'purple-600',
      hover_color: 'purple-700',
      border_color: 'purple-200',
      bg_color: 'purple-50',
      text_color: 'purple-600'
    }
  },

  count: {
    type: 'count',
    title: 'Stock Count',
    color: 'yellow',
    locations: ['destination'],
    validation_rules: ['positive_quantity'],
    quantity_behavior: 'positive',
    description: 'Record stock count for inventory verification',
    ui_behavior: {
      show_current_stock: true,
      allow_negative_quantity: false,
      require_location_selection: true,
      default_quantity: 0,
      quantity_step: 1,
      quantity_min: 0,
      show_count_difference: true
    },
    api_endpoints: {
      search: '/teams/{teamId}/items/search',
      transaction: '/teams/{teamId}/transactions'
    },
    validation_messages: {
      no_location: 'Please select a location',
      no_items: 'Please add items and quantities',
      invalid_quantity: 'Count must be zero or positive'
    },
    css_classes: {
      primary_color: 'yellow-600',
      hover_color: 'yellow-700',
      border_color: 'yellow-200',
      bg_color: 'yellow-50',
      text_color: 'yellow-600'
    }
  }
}

/**
 * Get configuration for a specific transaction type
 * @param {string} type - Transaction type
 * @param {string} teamId - Team ID for API endpoints
 * @returns {object} Configuration object with interpolated API endpoints
 */
export function getTransactionConfig(type, teamId = null) {
  const config = TRANSACTION_CONFIGS[type]
  if (!config) {
    throw new Error(`Unknown transaction type: ${type}`)
  }

  // Create a deep copy to avoid modifying the original
  const configCopy = JSON.parse(JSON.stringify(config))

  // Interpolate team ID in API endpoints if provided
  if (teamId) {
    Object.keys(configCopy.api_endpoints).forEach(key => {
      configCopy.api_endpoints[key] = configCopy.api_endpoints[key].replace('{teamId}', teamId)
    })
  }

  return configCopy
}

/**
 * Get all available transaction types
 * @returns {string[]} Array of transaction type names
 */
export function getAvailableTransactionTypes() {
  return Object.keys(TRANSACTION_CONFIGS)
}

/**
 * Check if a transaction type is valid
 * @param {string} type - Transaction type to validate
 * @returns {boolean} True if valid
 */
export function isValidTransactionType(type) {
  return TRANSACTION_CONFIGS.hasOwnProperty(type)
}

/**
 * Get validation rules for a transaction type
 * @param {string} type - Transaction type
 * @returns {string[]} Array of validation rule names
 */
export function getValidationRules(type) {
  const config = TRANSACTION_CONFIGS[type]
  return config ? config.validation_rules : []
}

/**
 * Get UI behavior configuration for a transaction type
 * @param {string} type - Transaction type
 * @returns {object} UI behavior configuration
 */
export function getUIBehavior(type) {
  const config = TRANSACTION_CONFIGS[type]
  return config ? config.ui_behavior : {}
}

/**
 * Get CSS classes for a transaction type
 * @param {string} type - Transaction type
 * @returns {object} CSS class configuration
 */
export function getCSSClasses(type) {
  const config = TRANSACTION_CONFIGS[type]
  return config ? config.css_classes : {}
}

/**
 * Get validation messages for a transaction type
 * @param {string} type - Transaction type
 * @returns {object} Validation message configuration
 */
export function getValidationMessages(type) {
  const config = TRANSACTION_CONFIGS[type]
  return config ? config.validation_messages : {}
}

/**
 * Create a configuration object suitable for passing to Stimulus controllers
 * @param {string} type - Transaction type
 * @param {string} teamId - Team ID
 * @param {object} overrides - Optional configuration overrides
 * @returns {object} Configuration object for Stimulus controller
 */
export function createStimulusConfig(type, teamId, overrides = {}) {
  const baseConfig = getTransactionConfig(type, teamId)
  
  // Merge with any overrides
  const config = {
    ...baseConfig,
    ...overrides,
    ui_behavior: {
      ...baseConfig.ui_behavior,
      ...(overrides.ui_behavior || {})
    },
    api_endpoints: {
      ...baseConfig.api_endpoints,
      ...(overrides.api_endpoints || {})
    },
    validation_messages: {
      ...baseConfig.validation_messages,
      ...(overrides.validation_messages || {})
    },
    css_classes: {
      ...baseConfig.css_classes,
      ...(overrides.css_classes || {})
    }
  }

  return config
}

export default {
  TRANSACTION_CONFIGS,
  getTransactionConfig,
  getAvailableTransactionTypes,
  isValidTransactionType,
  getValidationRules,
  getUIBehavior,
  getCSSClasses,
  getValidationMessages,
  createStimulusConfig
}

File: app/models/concerns/transaction_config.rb
# frozen_string_literal: true

module TransactionConfig
  extend ActiveSupport::Concern

  # Configuration constants for all transaction types
  TRANSACTION_CONFIGS = {
    stock_in: {
      title: 'Stock In',
      color: 'green',
      locations: [:destination],
      validation_rules: [:positive_quantity],
      quantity_behavior: :positive,
      description: 'Add items to inventory at a specific location'
    }.freeze,
    
    stock_out: {
      title: 'Stock Out',
      color: 'red', 
      locations: [:source],
      validation_rules: [:negative_quantity, :stock_availability],
      quantity_behavior: :negative,
      description: 'Remove items from inventory at a specific location'
    }.freeze,
    
    adjust: {
      title: 'Adjust Stock',
      color: 'blue',
      locations: [:destination],
      validation_rules: [:adjustment_calculation],
      quantity_behavior: :adjustment,
      description: 'Adjust item quantities to match actual counts'
    }.freeze,
    
    move: {
      title: 'Move Stock',
      color: 'purple',
      locations: [:source, :destination],
      validation_rules: [:positive_quantity, :stock_availability],
      quantity_behavior: :positive,
      description: 'Transfer items between locations'
    }.freeze,
    
    count: {
      title: 'Stock Count',
      color: 'yellow',
      locations: [:destination],
      validation_rules: [:positive_quantity],
      quantity_behavior: :positive,
      description: 'Record stock count for inventory verification'
    }.freeze
  }.freeze

  class_methods do
    # Get configuration for a specific transaction type
    def transaction_config(type)
      config = TRANSACTION_CONFIGS[type.to_sym]
      raise ArgumentError, "Unknown transaction type: #{type}" unless config
      
      TransactionConfigStruct.new(type.to_sym, config)
    end

    # Get all available transaction types
    def available_transaction_types
      TRANSACTION_CONFIGS.keys
    end

    # Check if a transaction type is valid
    def valid_transaction_type?(type)
      TRANSACTION_CONFIGS.key?(type.to_sym)
    end
  end

  included do
    # Instance method to get configuration for this transaction's type
    def config
      return nil unless transaction_type.present?
      self.class.transaction_config(transaction_type)
    end
  end

  # Struct-like class to provide convenient access to configuration data
  class TransactionConfigStruct
    attr_reader :type, :title, :color, :locations, :validation_rules, :quantity_behavior, :description

    def initialize(type, config_hash)
      @type = type
      @title = config_hash[:title]
      @color = config_hash[:color]
      @locations = config_hash[:locations]
      @validation_rules = config_hash[:validation_rules]
      @quantity_behavior = config_hash[:quantity_behavior]
      @description = config_hash[:description]
    end

    # Location requirement helper methods
    def requires_source_location?
      locations.include?(:source)
    end

    def requires_destination_location?
      locations.include?(:destination)
    end

    def requires_both_locations?
      requires_source_location? && requires_destination_location?
    end

    def requires_single_location?
      locations.length == 1
    end

    # Validation rule helper methods
    def requires_positive_quantity?
      validation_rules.include?(:positive_quantity)
    end

    def requires_negative_quantity?
      validation_rules.include?(:negative_quantity)
    end

    def requires_stock_availability_check?
      validation_rules.include?(:stock_availability)
    end

    def requires_adjustment_calculation?
      validation_rules.include?(:adjustment_calculation)
    end

    # Quantity behavior helper methods
    def quantity_should_be_positive?
      quantity_behavior == :positive
    end

    def quantity_should_be_negative?
      quantity_behavior == :negative
    end

    def quantity_is_adjustment?
      quantity_behavior == :adjustment
    end

    # UI helper methods
    def css_color_class
      "text-#{color}-600"
    end

    def bg_color_class
      "bg-#{color}-50"
    end

    def border_color_class
      "border-#{color}-200"
    end

    def button_color_class
      "bg-#{color}-600 hover:bg-#{color}-700"
    end

    # Convert to hash for JSON serialization
    def to_h
      {
        type: type,
        title: title,
        color: color,
        locations: locations,
        validation_rules: validation_rules,
        quantity_behavior: quantity_behavior,
        description: description
      }
    end

    def to_json(*args)
      to_h.to_json(*args)
    end
  end
end